app/response_processor.py

```diff
@@ -9,130 +8,101 @@
-        except KeyError:
-            # No secret defined; message is plaintext
-            pass
-
-        decrypted_json = loads(message)
-        if "metadata" not in decrypted_json:
-            return False
-
-        metadata = decrypted_json['metadata']
-        self.logger = self.logger.bind(user_id=metadata['user_id'], ru_ref=metadata['ru_ref'])
-
-        if 'tx_id' in decrypted_json:
-            self.tx_id = decrypted_json['tx_id']
-            self.logger = self.logger.bind(tx_id=self.tx_id)
-
-        receipt_ok = self.send_receipt(decrypted_json)
-        if not receipt_ok:
-            return False
-        else:
-            return True
-
-    def send_receipt(self, decrypted_json):
-        if self.skip_receipt:
-            self.logger.debug("Skipping sending receipt to RRM")
-            return True
-        else:
-            self.logger.debug("Sending receipt to RRM")
-
-        endpoint = receipt.get_receipt_endpoint(decrypted_json)
-        if endpoint is None:
-            return False
+        if "metadata" not in decrypted:
+            raise BadMessageError("Missing metadata")
+        return
 
-        xml = receipt.get_receipt_xml(decrypted_json)
+    def _encode(self, decrypted):
+        xml = receipt.get_receipt_xml(decrypted)
         if xml is None:
-            return False
+            raise BadMessageError("Unable to generate xml from message")
+        return xml
 
-        headers = receipt.get_receipt_headers()
+    def _send_receipt(self, decrypted, xml):
+        endpoint = receipt.get_receipt_endpoint(decrypted)
+        if endpoint is None:
+            raise BadMessageError("Unable to determine delivery endpoint from message")
 
-        response = self.remote_call(
-            endpoint,
-            data=xml.encode("utf-8"),
-            headers=headers,
-            verify=False,
-            auth=(settings.RECEIPT_USER, settings.RECEIPT_PASS))
-        return self.response_ok(response)
+        headers = receipt.get_receipt_headers()
+        auth = (settings.RECEIPT_USER, settings.RECEIPT_PASS)
 
-    def remote_call(self, request_url, json=None, data=None, headers=None, verify=True, auth=None):
         try:
-            self.logger.info("Calling service", request_url=request_url)
-            r = None
-
-            if data:
-                r = session.post(request_url, data=data, headers=headers, verify=verify, auth=auth)
-            else:
-                r = session.get(request_url, headers=headers, verify=verify, auth=auth)
+            self.logger.info("Calling service", request_url=endpoint)
+            res = session.post(endpoint, data=xml, headers=headers, verify=False, auth=auth)
 
-            return r
+            if res.status_code == 400:
+                self.logger.error("Receipt rejected by endpoint", request_url=res.url, status_code=400)
+                raise BadMessageError("Failure to send receipt")
```

Comment:
Would not use the same event text for BadMessage and the following RetryableError as they have different effects? Also logging out the tx_id and/or xml details would be required to make use of this log line.